### PR TITLE
CR-02: Add initial Dagger 2 configuration

### LINKED ISSUE #3: CR-02: Add initial Dagger 2 configuration
- Implement custom app class
- Implement Dagger 2 application component
- Implement Dagger 2 activity component
- Implement scopes
- Create an initial dependency graph

### PR DESCRIPTION
**Pull request by tartarJR**

### CODE REVIEW

File: app/src/main/java/com/tatar/currencify/presentation/viewmodel/ViewModelFactory.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        if (provider == null) {
+            for ((key, value) in providerMap) {
+                if (modelClass.isAssignableFrom(key)) {
+                    provider = value
+                    break
+                }
+            }
+        }
+
+        if (provider == null) {
```
Comment (flyingBear66) [inline app/src/main/java/com/tatar/currencify/presentation/viewmodel/ViewModelFactory.kt:26]: Why not else here? so you would get rid of an if that checks same thing?

Comment (flyingBear66) [reply app/src/main/java/com/tatar/currencify/presentation/viewmodel/ViewModelFactory.kt:26]: Also, I'd prefer to put this throw in the top of function as a quick return/exit. What was the reason you put here after?

Comment (flyingBear66) [reply app/src/main/java/com/tatar/currencify/presentation/viewmodel/ViewModelFactory.kt:26]: In Swift we have one thing called guard so you would do something similar like below. So, you wouldn't need any indentation, basically code would be clearer. Does Kotlin have something similar?

**Review by flyingBear66 — COMMENTED** [review]
Neat code! Just have some one doubt on one case.

File: app/src/main/java/com/tatar/currencify/presentation/viewmodel/ViewModelFactory.kt (hunk 1 above)
Comment (tartarJR) [reply app/src/main/java/com/tatar/currencify/presentation/viewmodel/ViewModelFactory.kt:26]: This class is a weird one. The reason for the weirdness is injecting Android's `ViewModel`s via Dagger 2. It is a long story but the story is the reason for implementing this class in that way.

**Review by flyingBear66 — APPROVED** [review]
Apart from weird Dagger Implementation that we can't resolve, LGTM!

**Pull request merged**